ipn/ipnlocal/state_test.go

```diff
@@ -0,0 +1,778 @@
+	// (and when it gets close to expiring)
+
+	// The user changes their preference to !WantRunning.
+	t.Logf("\n\nWantRunning -> false")
+	notifies.expect(2)
+	b.EditPrefs(&ipn.MaskedPrefs{
+		WantRunningSet: true,
+		Prefs:          ipn.Prefs{WantRunning: false},
+	})
+	{
+		nn := notifies.drain(2)
+		assert.Equal([]string{"pause"}, cc.getCalls())
+		// BUG: I would expect Prefs to change first, and state after.
+		assert.NotNil(nn[0].State)
+		assert.NotNil(nn[1].Prefs)
+		assert.Equal(ipn.Stopped, *nn[0].State)
+	}
+
+	// The user changes their preference to WantRunning after all.
+	t.Logf("\n\nWantRunning -> true")
+	notifies.expect(2)
+	b.EditPrefs(&ipn.MaskedPrefs{
+		WantRunningSet: true,
+		Prefs:          ipn.Prefs{WantRunning: true},
+	})
+	{
+		nn := notifies.drain(2)
+		// BUG: UpdateEndpoints isn't needed here.
+		// BUG: Login isn't needed here. We never logged out.
+		assert.Equal([]string{"unpause", "UpdateEndpoints", "Login"}, cc.getCalls())
+		// BUG: I would expect Prefs to change first, and state after.
+		assert.NotNil(nn[0].State)
+		assert.NotNil(nn[1].Prefs)
+		assert.Equal(ipn.Starting, *nn[0].State)
+	}
+
+	// Test the fast-path frontend reconnection.
+	// This one is very finicky, so we have to force State==Running.
+	// TODO: actually get to State==Running, rather than cheating.
+	t.Logf("\n\nFastpath Start()")
+	notifies.expect(1)
+	b.state = ipn.Running
+	assert.Nil(b.Start(ipn.Options{
+		StateKey: ipn.GlobalDaemonStateKey,
+	}))
+	{
+		nn := notifies.drain(1)
+		assert.Equal([]string{}, cc.getCalls())
+		assert.NotNil(nn[0].State)
+		assert.NotNil(nn[0].LoginFinished)
+		assert.NotNil(nn[0].NetMap)
+		// BUG: Prefs should be sent too, or the UI could end up in
+		// a bad state. (iOS, the only current user of this feature,
+		// probably wouldn't notice because it happens to not display
+		// any prefs. Maybe exit nodes will look weird?)
+	}
+
+	// undo the state hack above.
+	b.state = ipn.Starting
+
+	// User wants to logout.
+	t.Logf("\n\nLogout (async)")
+	notifies.expect(1)
+	b.Logout()
+	{
+		nn := notifies.drain(1)
+		assert.Equal([]string{"StartLogout"}, cc.getCalls())
+		assert.NotNil(nn[0].Prefs)
+		// NOTE: WantRunning is false here, which is right. But
+		// it changes to true in subsequent logout attempts, which is
+		// wrong in those attempts.
+		assert.Equal(false, nn[0].Prefs.WantRunning)
+	}
+
+	// Let's make the logout succeed.
+	notifies.expect(1)
+	cc.setAuthBlocked(true)
+	cc.send(nil, "", false, nil)
+	{
+		nn := notifies.drain(1)
+		// BUG: now is not the time to unpause.
+		assert.Equal([]string{"unpause"}, cc.getCalls())
+		assert.NotNil(nn[0].State)
+		assert.Equal(ipn.NeedsLogin, *nn[0].State)
+	}
+
+	// A second logout should do nothing, since the prefs haven't changed.
+	t.Logf("\n\nLogout2 (async)")
+	notifies.expect(0)
+	b.Logout()
+	{
+		notifies.drain(0)
+		assert.Equal([]string{"StartLogout"}, cc.getCalls())
+	}
+
+	// Let's acknowledge the second logout too.
```

Comment:
Should controlclient even get told about a second logout? This feels like ipn's doing a lot of message passing between components to implement "I know this won't do anything, and I have correspondingly done nothing" (bugs aside, that is)

Reply:
The comment a few lines up is trying to say no, it shouldn't propagate the second Logout request since it knows it's already logged out. I should add a BUG line here to make that clear.